scipy/integrate/_quadrature.py

```diff
@@ -1051,65 +1045,82 @@ def cumulative_simpson(y, *, x=None, dx=1.0, axis=-1, initial=None):
     >>> ax.grid()
     >>> plt.show()
 
+    The output of `cumulative_simpson` is similar to that of iteratively
+    calling `simpson` with successively larger integration intervals, but
+    not identical.
+
+    >>> def cumulative_simpson_reference(y, x):
+    ...     return np.asarray([integrate.simpson(y[:i], x=x[:i])
+    ...                        for i in range(2, len(y) + 1)])
+    >>>
+    >>> rng = np.random.default_rng(354673834679465)
+    >>> x, y = rng.random(size=(2, 10))
+    >>> x.sort()
+    >>>
+    >>> res = integrate.cumulative_simpson(y, x=x)
+    >>> ref = cumulative_simpson_reference(y, x)
+    >>> equal = np.abs(res - ref) < 1e-15
+    >>> equal  # not equal when `simpson` has even number of subintervals
+    array([False,  True, False,  True, False,  True, False,  True,  True])
+
+    This is expected: because `cumulative_simpson` has access to more
+    information than `simpson`, it can typically produce more accurate
+    estimates of the underlying integral over subintervals.
+
     """
-    y = _convert_input_to_float_array(y)
+    y = _ensure_float_array(y)
 
-    # validate y and axis
-    if axis < -1 or axis >= y.ndim:
-        raise ValueError("If given, `axis` must exist in the shape of `y`.")
+    # validate `axis` and standardize to work along the last axis
+    original_shape = y.shape
+    try:
+        y = np.moveaxis(y, axis, -1)
+    except np.exceptions.AxisError as e:
+        message = f"`axis={axis}` is not valid for `y` with `y.ndim={y.ndim}`."
+        raise ValueError(message) from e
 
-    if y.shape[axis] < 3:
+    if y.shape[-1] < 3:
         raise ValueError(
             "At least 3 points are required along the axis of integration "
             "to use the composite Simpson's method."
         )
 
     if x is not None:
-        x = _convert_input_to_float_array(x)
-        if x.ndim == 1 and y.shape[axis] == x.shape[0]:
-            x_shape = [1] * y.ndim
-            x_shape[axis] = -1
-            x = x.reshape(tuple(x_shape))
-        elif x.shape != y.shape:
-            raise ValueError(
-                "If given, shape of `x` must be the same as `y` or 1-D with "
-                "the same length as `y` along `axis`."
-            )
-        res = _cumulative_simpson_unequal_intervals(y, x, axis=axis)
+        x = _ensure_float_array(x)
+        message = ("If given, shape of `x` must be the same as `y` or 1-D with "
+                   "the same length as `y` along `axis`.")
+        if not (x.shape == original_shape
+                or (x.ndim == 1 and len(x) == original_shape[axis])):
+            raise ValueError(message)
+
+        x = np.broadcast_to(x, y.shape) if x.ndim == 1 else np.moveaxis(x, axis, -1)
+        res = _cumulative_simpson_unequal_intervals(y, x)
```

Comment:
Here and below, I've tried to make the code a bit more concise, and I've relied on NumPy to do the messy stuff instead of ad hoc code. Especially below, I prefer the more linear structure -  get the input validation out of the way, then do the same thing for any input instead of a three-way branch structure.